### PR TITLE
refactor(core): strong type keyvalue differs

### PR DESCRIPTION
**Pull request by alfaproject**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?**
KeyValueDiffers and friends are not strongly typed

**What is the new behavior?**
KeyValueDiffers and friends are strongly typed

**Does this PR introduce a breaking change?**

```
[x] Yes
[ ] No
```

### CODE REVIEW

File: modules/@angular/common/src/directives/ng_class.ts [hunk 1]
```diff
@@ -40,12 +40,13 @@ import {isPresent} from '../facade/lang';
 @Directive({selector: '[ngClass]'})
 export class NgClass implements DoCheck {
   private _iterableDiffer: IterableDiffer;
-  private _keyValueDiffer: KeyValueDiffer;
+  private _keyValueDiffer: KeyValueDiffer<string[]|Set<string>|{[klass: string]: boolean}>;
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_class.ts]: revert the boolean stuff (JS accepts any if ('true'))

Comment (alfaproject) [inline modules/@angular/common/src/directives/ng_class.ts]: I've just reverted it

**Comment by vicb** [comment]
API break -> can not merge before 3.0

**Comment by alfaproject** [comment]
Pity ):

**Comment by vicb** [comment]
BTW thanks for all your PRs... 
Trying to find some spare time to go over them.

File: tools/public_api_guard/common/index.d.ts [hunk 2]
```diff
@@ -113,7 +113,9 @@ export declare class NgClass implements DoCheck {
     ngClass: string | string[] | Set<string> | {
         [klass: string]: any;
     };
-    constructor(_iterableDiffers: IterableDiffers, _keyValueDiffers: KeyValueDiffers, _ngEl: ElementRef, _renderer: Renderer);
+    constructor(_iterableDiffers: IterableDiffers, _keyValueDiffers: KeyValueDiffers<string[] | Set<string> | {
```
Comment (mhevery) [inline tools/public_api_guard/common/index.d.ts]: KeyValue differ can not have `string[]` or `Set<string>` as its type. This does not seem right.

**Review by mhevery — CHANGES_REQUESTED** [review]

File: modules/@angular/core/src/change_detection/differs/keyvalue_differs.ts [hunk 3]
```diff
@@ -17,36 +15,35 @@ import {ChangeDetectorRef} from '../change_detector_ref';
...
+  create(cdRef: ChangeDetectorRef): KeyValueDiffer<T>;
 }
 
 /**
  * A repository of different Map diffing strategies used by NgClass, NgStyle, and others.
  * @stable
  */
-export class KeyValueDiffers {
-  constructor(public factories: KeyValueDifferFactory[]) {}
+export class KeyValueDiffers<T> {
```
Comment (mhevery) [inline modules/@angular/core/src/change_detection/differs/keyvalue_differs.ts]: `KeyValueDiffers` can not be parameterized as it holds differs for many different parameters.

**Review by mhevery — COMMENTED** [review]
I will take over this PR

**Comment by mhevery** [comment]
Closing this in favor of https://github.com/angular/angular/pull/13382

**Pull request closed without merging**